app/assets/javascripts/cart.js

```diff
@@ -0,0 +1,163 @@
+function shoppingCart(cartName) {
+    this.cartName = cartName;
+    this.items = [];
+    this.clearCart = false;
+
+    this.loadItems();
+    var self = this;
+    $(window).unload(function () {
+        if (self.clearCart) {
+            self.clearItems();
+        }
+        self.saveItems();
+        self.clearCart = false;
+    });
+}
+
+shoppingCart.prototype.loadItems = function () {
+    var items = localStorage !== null ? localStorage[this.cartName + "_items"] : null;
+    if (items !== null && JSON !== null) {
+        try {
+            items = JSON.parse(items);
+            for (var i = 0; i < items.length; i++) {
+                var item = items[i];
+                if (item.code !== null && item.name !== null && item.price
+                        !== null && item.quantity != null &&
+                        item.image != null) {
+                    item = new cartItem(
+                        item.code,
+                        item.name,
+                        item.price,
+                        item.quantity,
+                        item.image
+                    );
+                    this.items.push(item);
+                }
+            }
+        }
+        catch (err) {
+            // ignore errors while loading...
+        }
+    }
+};
+
+shoppingCart.prototype.saveItems = function () {
+    if (localStorage !== null && JSON !== null) {
+        localStorage[this.cartName + "_items"] = JSON.stringify(this.items);
+    }
+};
+
+// adds an item to the cart
+shoppingCart.prototype.addItem = function (code, name, price, quantity, image) {
+    quantity = this.toNumber(quantity);
+    if (quantity !== 0) {
+
+        // update quantity for existing item
+        var found = false;
+        for (var i = 0; i < this.items.length && !found; i++) {
+            var item = this.items[i];
+            if (item.code == code) {
+                found = true;
+                item.quantity = this.toNumber(item.quantity + quantity);
+                if (item.quantity <= 0) {
+                    this.items.splice(i, 1);
+                }
+            }
+        }
+
+        // new item, add now
+        if (!found) {
+            item = new cartItem(code, name, price, quantity, image);
```

Comment:
'item' used out of scope.<br>A constructor name should start with an uppercase letter.